reps/2023-08-18-serve-java-dag-api.md

```diff
@@ -0,0 +1,266 @@
+## Summary
+### General Motivation
+Compared to the latest Serve API in Python, Ray Serve Java lacks a major version update. Currently, the main form of the Serve API is the DAG API, which involves networking deployments through binding and deploying the graph. For example:
+```python
+import ray
+from ray import serve
+from ray.serve.dag import InputNode
+from ray.serve.drivers import DAGDriver
+
+
+@serve.deployment
+def preprocess(inp: int) -> int:
+    return inp + 1
+
+
+@serve.deployment
+class Model:
+    def __init__(self, increment: int):
+        self.increment = increment
+
+    def predict(self, inp: int) -> int:
+        return inp + self.increment
+
+
+with InputNode() as inp:
+    model = Model.bind(increment=2)
+    output = model.predict.bind(preprocess.bind(inp))
+    serve_dag = DAGDriver.bind(output)
+
+handle = serve.run(serve_dag)
+assert ray.get(handle.predict.remote(1)) == 4
```

Comment:
@liuyang-my we are actually discouraging the use of the "call graph" api using `with InputNode` and instead encouraging users to use the deployment handle interface directly to perform composition. For this example it'd look like:

```python
@serve.deployment
def preprocess(inp: int) -> int:
    return inp + 1


@serve.deployment
class Model:
    def __init__(self, preprocess_handle: RayServeHandle, increment: int):
        self.preprocess_handle: DeploymentHandle = preprocess_handle.options(use_new_handle_api=True)
        self.increment = increment

    async def predict(self, inp: int): -> int:
        preprocessed = await self.preprocess_handle.remote(inp)
        return inp + self.increment

app = Model.bind(preprocess.bind(), increment=2)

handle = serve.run(app)
assert ray.get(handle.predict.remote(1)) == 4
```

I think we can match this API very closely in Java

Reply:
It has been adjusted.